engine/src/main/java/org/terasology/rendering/shader/ShaderParametersBase.java

```diff
@@ -43,27 +45,29 @@ public void applyParameters(Material program) {
         program.setFloat("viewingDistance", CoreRegistry.get(Config.class).getRendering().getViewDistance().getChunkDistance().x * 8.0f);
 
         WorldRenderer worldRenderer = CoreRegistry.get(WorldRenderer.class);
-        WorldProvider worldProvider = CoreRegistry.get(WorldProvider.class);
+        BackdropProvider backdropProvider = CoreRegistry.get(BackdropProvider.class);
 
-        if (worldRenderer != null) {
-            program.setFloat("daylight", worldRenderer.getDaylight(), true);
+        if (worldRenderer != null && backdropProvider != null) {
```

Comment:
There's a mixture of stuff in the ShaderParameters classes. 

Some of it is one-off initialisation - this can just be done as part of the material (and thus in json). For instance, there is nothing in ShaderParameterBlock (excluding inherited values) that needs to be done at all every frame - everything is constants and the texture binding can be done by the material.

Some of it is applying "environmental" values. This is where the shader could declare a meaning (semantic) for a variable, and the shader system would know what values to inject in those variables. For instance, a shader could have a variable declared like ```uniform mat3 normalMatrix : NORMAL_MATRIX``` to indicate that it should be populated with the normal matrix for the current camera. Obviously this is not a glsl feature, so it would require some work on our part. The upshot is you drop the idea of custom shader parameter injection classes, and instead provide a range of features that shaders can ask for as part of their definition. We would presumably have a class for each of these semantics, but it would be a much more flexible model than the current one.

A small part of it is very specific changing values that are very shader specific. I would still drop ShaderParameters for this, and instead just have a system update the material as needed. To aid this I would have support for material instances - these are copies of a material that can have separate variable values (I'm planning this for gestalt-asset-core). These would still link back to their asset of origin for persistence.

I would disagree on there being a lot of logic in the ShaderParameter classes - it is all just copying values into materials from various sources.